language/move-compiler/src/naming/ast.rs

```diff
@@ -360,15 +367,17 @@ impl BuiltinTypeName_ {
             B::Address | B::U8 | B::U64 | B::U128 | B::Bool => AbilitySet::primitives(loc),
             B::Signer => AbilitySet::signer(loc),
             B::Vector => AbilitySet::collection(loc),
+            B::Fun => AbilitySet::functions(),
         }
     }
 
-    pub fn tparam_constraints(&self, _loc: Loc) -> Vec<AbilitySet> {
+    pub fn tparam_constraints(&self, _loc: Loc, arg_count: usize) -> Vec<AbilitySet> {
         use BuiltinTypeName_ as B;
         // Match here to make sure this function is fixed when collections are added
         match self {
             B::Address | B::Signer | B::U8 | B::U64 | B::U128 | B::Bool => vec![],
             B::Vector => vec![AbilitySet::empty()],
+            B::Fun => (0..arg_count).map(|_| AbilitySet::empty()).collect(),
```

Comment:
As mentioned in the PR description, it is an intentional design approach right now to make this a (builtin) type constructor instead of a new type variant. This keeps the footprint of this PR significantly lower (I know because I first added a new type variant and saw how much this impacts existing code and how much changes would be needed). 

I think it is also conceptually cleaner, and tuples could be represented as type constructors as well. Those things behave nearly the same as struct type constructors, and having a smaller formal language is in my experience preferable. The difference is only in co/contra-variance. I have frankly not taken too much care of the subtype thing, because I did not think there are subtypes (?), but I can fix the type unification to do this right. Conceptually, for unification one would ask the type constructor whether it is co/contra-variant in the ith of n arguments.

I can change this back to having a new type variant instead of the constructor, but I wanted to check first how strong you feel about it. Lets first clarify this before addressing other concerns here,